### DISCUSSION TITLE
HandleAsync not supported

### DISCUSSION THREAD

**Discussion by pedrollanes** [canton7/Stylet#189] [Ideas]
When you use EventAggregator you can't handle async. I think that it's important because in the handler, if you want to call to async method you are force to use GetAwaiter, and in some cases I have deathlocks. Are you going to develop HandleAsync for eventAggregator?

**Comment by canton7** [comment]
> Are you going to develop HandleAsync for eventAggregator?

I've got no plans to, as noone has asked for it 🙂

You're welcome to contribute this feature: it should be relatively straightforward, once you've decided on a design.

The open questions are:

1. What should happen to the `Task` which is returned? Should `EventAggregator` wait until the `Task` returned from one handler is completed before invoking the next handler?
2. What should happen if the `Task` returned from a handler completes with an exception: should that be rethrown to the caller, or something else?

Be careful of cases where you need `EventAggregator` to gain `PublishAsync` methods, and you need to start worrying about what should happen if someone calls `Publish` when there are async handlers.

Note that your handler methods can be `async void`: they're essentially the same as event handlers, so it makes sense that they do their own error-handling, and don't report status to their caller. This is probably the most straightforward approach.

**Reply by pedrollanes** [reply]
Nowadays I have no time for it, but in the future maybe I'll try. Thanks for you reply

**Comment by alexhelms** [comment]
Moving over from #238, I didn't see this one.

I'd be willing to work on this if we can agree on some answers to the questions asked by @canton7.

I think we should follow MediatR's way of doing notifications. I found a thread here talking about notifications and exceptions and I agree with this post: https://github.com/jbogard/MediatR/issues/237#issuecomment-366911171

So just to clarify:

1. _What should happen to the Task which is returned?_ I propose we await it.
2. _What should happen if the Task returned from a handler completes with an exception?_ I propose we let it bubble up because events should a) have their own error handling, or b) ensure they don't throw exceptions.

Thoughts?

**Reply by canton7** [reply]
> I propose we await it.

Take an event with multiple subscribers. The EventAggregator invokes the first and gets a Task back. Does the EventAggregator wait until the Task is complete before invoking the next one? Some of your subscribes could be very delayed! You also have the problem that events can start arriving out-of-order, which is going to be bad.

>  I propose we let it bubble up 

Bubble up to where? We can't bubble it up to the method which published the event, because publishing isn't an async operation, so the call to `Publish` returned long before an Exception ever appeared. We could re-throw it on the dispatcher/threadpool and crash the app, or we could swallow it, or we could introduce an entirely separate `AsyncEventAggregator` which has an async `PublishAsync` method.

If your view is that the subscribers shouldn't thow exceptions, and the EventAggregator shouldn't wait for one subscriber to finish to completion before calling the next, then just make the subscriber methods `async void` -- it's what it's there for.

**Reply by alexhelms** [reply]
There are many ways to handle multiple events but comparing to MediatR, they provide a way to have custom publish strategies. I don't think Stylet would need publishing strategies for the first stab at this but it is an option to consider. I think the default way is what somebody would typically want for a MVVM event aggregator. A new PublishAsync function that behaves just as the current synchronous one but is compatible with TPL -- sequentially await each event handler while collecting any exceptions and throwing an aggregate exception if there are any. Take a look at [this example](https://github.com/jbogard/MediatR/blob/50af3de7fa89d56067952581179fa22a1b66de0b/samples/MediatR.Examples.PublishStrategies/Publisher.cs#L131).

I don't understand when you say "publishing isn't an async operation, so the call to `Publish` returned long before an Exception appeaered". What I'm describing is that the caller would `await aggregator.PublishAsync(new Ping())` and saying there are 3 subscribers, each would be awaited in the order in which they subscribed, and then `PublishAsync` would complete and the original publisher moves on. In this scenario the publisher is still there and if there are exceptions in any of the handlers `PublishAsync` would have throw an `AggregateException` that contains the inner exceptions.

I suggested that event subscribers shouldn't throw exceptions because events are supposed to do as little work as possible. I would consider an event handler intentionally throwing an exception to be a bug and shouldn't be part of the design. I thought the thread I linked in my last post demonstrated that pretty well. I also think the C# community agrees that `async void` should be avoided, Stephen Cleary has said lots on that subject. By providing a mechanism to sequentially await each handler and potentially return an `AggregateException` is the better approach. Ideally there should be no exception, but if there is the caller should be the one to deal with it and not the global dispatcher unhandled exception handler.

**Reply by canton7** [reply]
We already have a Publish method. Any proposal needs to work with it, first and foremost. From my original post:

> Be careful of cases where you need EventAggregator to gain PublishAsync methods, and you need to start worrying about what should happen if someone calls Publish when there are async handlers.

**Reply by canton7** [reply]
I did actually discuss the question of async void with Stephen on SO some years back, and he clarified that "avoid async void" was simplified advice for newbies, and that it does have its place when used correctly.

**Reply by canton7** [reply]
I suspect the only realistic way forwards may be a separate IAsyncEventAggregator with async Publish methods and async/sync subscribe methods.

**Reply by alexhelms** [reply]
Yea, that seems to be the cleanest approach. I assume that users would have to choose to use the sync or async aggregator since they would be separate things? Or should the existing implementation handle both interfaces?

**Reply by canton7** [reply]
I suspect there probably won't be that much in common: the signatures of most of the methods, locking strategy, etc etc is all going to have to change. Perhaps a base/helper class if it helps.

I do feel strongly that the publish method shouldn't wait for the first subscriber to completely finish before invoking the second: a single mis-behaving subscriber will block all subsequent subscribers, and it opens up the door for subscribers to receive events out-of-order. E.g:

```cs
class Subscriber1 : IAsyncHandle<Event1>
{
    public Task Handle(Event1 e1) => Task.Delay(1000):
}

class Subscriber2 : IAsyncHandle<Event1>, IAsyncHandle<Event2>
{
    public async Task Handle(Event1 e1) => Console.WriteLine("E1");
    public async Task Handle(Event2 e2) => Console.WriteLine("E2");
}

var t1 = aggregator.PublishAsync(new Event1());
await aggregator.PublishAsync(new Event2());
```

... might print "E2 E1" which is wrong on all counts.

It also makes thread-safety a lot harder: if `Subscriber2` unsubscribes itself while the aggregator is waiting for `Subscriber1` to finish, you need to be careful it isn't subsequently invoked.

In order words, I think `Publish` needs to effectively do `await Task.WhenAll(subscribers.Select(x => x.PublishAsync(event))`.

**Reply by canton7** [reply]
Another Q: should the publish method allow cancellation? And if so should that be handled by passing a CT to subscribers?

**Reply by alexhelms** [reply]
I agree but I also think sequentially awaiting each handler is also useful. I guess this is why MediatR has publish strategies 😄 I'm fine with `Task.WhenAll` for awaiting when all handlers are complete. 

In your example `t1` is not being awaited. That tells me firing `Event1` is a "fire and forget" and you don't care about what happens to the task and getting "E2 E1" would be possible. I think that responsibility lies on the developer to not do that. They should have awaited the publish and that situation wouldn't have happened. If it were awaited the `AsyncEventAggregator` would have fired off all subscribed handlers and await them with a `Task.WhenAll` and returned when `Subscriber1` and `Subscriber2` were complete. Without the await the result would be determined by what the subscribers do.

Regarding cancellation, I think it would need to be passed through in order to stay consistent with normal async functions. The example below I think shows what you'd expect for using `WhenAll` and cancellation.

```c#
public class AsyncEventAggregator : IAsyncEventAggregator
{
    public Task PublishAsync(object message, CancellationToken token = default)
    {
        var tasks = new List<Task>();
        for (var handler in _handlers)
        {
            tasks.Add(handler.HandleAsync(message, token));
        }
        return Task.WhenAll(tasks);
    }
}

public class Subscriber1 : IHandleAsync<Event1>
{
    public Task HandleAsync(Event1 e, CancellationToken token)
    {
        return Task.CompletedTask;
    }
}
```

**Reply by canton7** [reply]
> In your example t1 is not being awaited. That tells me firing Event1 is a "fire and forget" and you don't care about what happens to the task and getting "E2 E1" would be possible. I think that responsibility lies on the developer to not do that. They should have awaited the publish and that situation wouldn't have happened.

`await Task.WhenAll(aggregator.Publish(new Event1()), aggregator.Publish(new Event2()))` then. The fact that E1 and E2 can arrive in the wrong order is going to catch people out. It's all very well saying "the developer should not do that", but you're not the one spending your evenings fielding the support requests!

Yeah I'm inclined to agree on the cancellation bit: it's annoying having to mandate that all subscribers handle cancellation, but someone's going to ask for it at some point and there's no way of putting it in after-the-fact.

**Reply by canton7** [reply]
Just to set expectations: any implementation of an async event aggregator needs to support the following before I can merge it:

1. Channels (using the same behaviour as the current non-async event aggregator)
2. Weak subscriptions
3. Publishing on different dispatchers
4. Thread-safety:
   a. Handlers can be added and removed at any point from any thread, including while an event is being fired.
   b. Two events can be fired simultaneously from different threads.
   c. This is going to be harder with the async version than with the sync version...
5. Handle non async subscribers 
6. We should do something sensible if someone subscribes to the non async aggregator with an async handler 
5. Full unit test coverage
6. Doc updates

I'm happy to do bits as well of course, but all those things need to be in place eventually.

**Comment by bradencross** [comment]
I am not sure if this is a related issue or not but I am running into a situation where the EventAggretator. Publish will indefinitely hang. Even if the subscribing handle is empty and not executing any code. Basically I am publishing an event in one viewmodel and consuming it in another. When this hang happens I can pause the debugger and can see it is stuck on the EventAggretor.Publish line. It does not happen every time but often enough that it is an issue. If anyone had some insight in what might be causing this would be helpful.

**Reply by canton7** [reply]
Turn off Just My Code and look at the call stack window. If you click through it, VS should download Stylet's source so you can see where exactly it's stuck. Without more info or code a self-contained example which reproduces the issue, it's hard to help I'm afraid!

**Reply by bradencross** [reply]
Fair enough! I have tried turning off Just my code but for some reason Stylet's source code isn't downloading properly (I will look into this). In the meantime I can now see that the lockup is in the WaitHelper method of the SynchronizationContext class. Specifically `WaitHandle.WaitMultipleIgnoringSyncContext(waitHandles, waitAll, millisecondsTimeout);` Thanks for the fast response, I will dig into this some more and get back to you. It appears that something in my code combined with EventAggregator.Publish is causing a deadlock.

**Reply by canton7** [reply]
That's weird, as `EventAggregator.Publish` doesn't do any threading: handlers are invoked synchronously on the calling thread.

The source is part of the symbols package, so you'll need "NuGet.org Symbol Server" enabled under Debug -> Options -> Symbols.

Please open a new discussion, with the full stack trace you have available, and the code which is invoking `EventAggregator.Publish`. Thanks!

**Comment by deralbert** [comment]
Hello @canton7. Is it known when will the asynchronous version of the EventAggregator be available? I understand that there is a lot of work behind this and I just want to know what the rough time frame for the introduction of the asynchronous EventAggregator is (and whether this introduction will take place at all).

Something similar happened to the Caliburn Micro. For a long time there were no asynchronous publishers and handlers and then sometime in 2018 they were introduced in such a way that you completely lost the possibility of using synchronous versions of `Publish` and `Handle` methods. It is also possible that the messages are processed in a different order than the order in which they were sent.

Taking into account what has already been discussed here with @alexhelms, the asynchronous version of EventAggregator next to the synchronous version would be a really nice addition to Stylet. I would like to help with the implementation, but I do not have a precise plan of which steps have to be taken first in which order and I have not found any open issues for AsyncEventAggregator.

**Reply by alexhelms** [reply]
I was hoping to do some work on async publishers for the past couple months but I haven't had time. So while I would still like to, I'm not able to commit to anything. Just wanted to throw that out there.

**Reply by canton7** [reply]
I'm not working on anything in the medium term, so no ETA.

Things like the event aggregator are pretty independent from the rest of Stylet: it's provided by Stylet only as a convenience. If you need an async event aggregator, I'd encourage you to write one which meets your needs. Use the existing synchronous one as a basis as much as you need. You could probably use C.M's if you need, since it's MIT. 

If you end up with something stable which meets your needs, please consider making a note of your learnings (why you had to change approach, for example), and contributing it to Stylet.

Thanks!
